packages/editor/src/extensions/code-block/code-block.ts

```diff
@@ -234,11 +234,34 @@ export const CodeBlock = Node.create<CodeBlockOptions>({
         },
       toggleCodeBlock:
         (attributes) =>
-        ({ commands }) => {
-          return commands.toggleNode(this.name, "paragraph", {
-            ...attributes,
-            id: createCodeblockId()
+        ({ commands, state }) => {
+          const { from, to } = state.selection;
+
+          let hasParagraph = false;
+
+          state.doc.nodesBetween(from, to, (node) => {
+            if (node.type.name === "paragraph") {
+              hasParagraph = true;
+            }
           });
+
+          if (hasParagraph) {
+            return commands.insertContent({
+              type: this.name,
+              attrs: {
+                ...attributes,
+                id: createCodeblockId()
+              },
+              content: [
+                {
+                  type: "text",
+                  text: state.doc.textBetween(from, to, "\n")
+                }
+              ]
+            });
```

Comment:
This can be simplified like so:

```ts
return commands.insertContent(
  this.type.create(
    { ...attributes, id: createCodeblockId() },
    state.schema.text(state.doc.textBetween(from, to, "\n"))
  )
);
```

However, I don't recommend depending on `insertContent` for replacing the selection. Instead we can directly use the `replaceSelectionWith` method on `Transaction`:

```ts
          const { from, to } = tr.selection;
          const text = tr.doc.textBetween(from, to, "\n");
          tr.replaceSelectionWith(
            this.type.create(
              { ...attributes, id: createCodeblockId() },
              state.schema.text(text)
            )
          );
```

Not to mention, that after replacing the node we also need to set the selection like so:

```ts
          commands.setTextSelection({ from, to: tr.mapping.map(to) });
```

---

This is all very rough so you might need to make adjustments.

Reply:
Sounds good, just changed this